pgvectorscale/src/util/chain.rs

```diff
@@ -0,0 +1,278 @@
+            let header = ChainItemHeader {
+                next: next_page.get_block_number(),
+            };
+            let header_bytes = rkyv::to_bytes::<_, 256>(&header).unwrap();
+            let data_size = current_page.get_free_space() - PG_SLOP_SIZE - CHAIN_HEADER_SIZE;
+            let chunk = &data[..data_size];
+            let combined = [header_bytes.as_slice(), chunk].concat();
+            let offset_number = current_page.add_item(combined.as_ref());
+            result.get_or_insert_with(|| {
+                ItemPointer::new(current_page.get_block_number(), offset_number)
+            });
+            current_page.commit();
+            self.stats.record_write();
+            current_page = next_page;
+            data = &data[data_size..];
+        }
+
+        // Write the last chunk of data.
+        let header = ChainItemHeader {
+            next: InvalidBlockNumber,
+        };
+        let header_bytes = rkyv::to_bytes::<_, 256>(&header).unwrap();
+        let combined = [header_bytes.as_slice(), data].concat();
+        let offset_number = current_page.add_item(combined.as_ref());
+        let result = result
+            .unwrap_or_else(|| ItemPointer::new(current_page.get_block_number(), offset_number));
+        self.current = current_page.get_block_number();
+        current_page.commit();
+        self.stats.record_write();
+
+        result
+    }
+}
+
+pub struct ChainTapeReader<'a, S: StatsNodeRead> {
+    page_type: PageType,
+    index: &'a PgRelation,
+    stats: &'a mut S,
+}
+
+impl<'a, S: StatsNodeRead> ChainTapeReader<'a, S> {
+    pub fn new(index: &'a PgRelation, page_type: PageType, stats: &'a mut S) -> Self {
+        assert!(page_type.is_chained());
+        Self {
+            page_type,
+            index,
+            stats,
+        }
+    }
+
+    pub fn read(&'a mut self, ip: ItemPointer) -> ChainedItemIterator<'a, S> {
+        ChainedItemIterator {
+            index: self.index,
+            ip,
+            page_type: self.page_type,
+            stats: self.stats,
+        }
+    }
+}
+
+pub struct ChainedItemIterator<'a, S: StatsNodeRead> {
+    index: &'a PgRelation,
+    ip: ItemPointer,
+    page_type: PageType,
+    stats: &'a mut S,
+}
+
+impl<'a, S: StatsNodeRead> Iterator for ChainedItemIterator<'a, S> {
+    type Item = ReadableBuffer<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.ip.block_number == InvalidBlockNumber {
+            return None;
+        }
+
+        unsafe {
+            let page = ReadablePage::read(self.index, self.ip.block_number);
+            self.stats.record_read();
+            assert!(page.get_type() == self.page_type);
+            let mut item = page.get_item_unchecked(self.ip.offset);
+            let slice = item.get_data_slice();
+            let header_slice = &slice[..CHAIN_HEADER_SIZE];
+            let header =
+                rkyv::from_bytes::<ChainItemHeader>(header_slice).expect("failed to read header");
```

Comment:
I don't think this is the zero-copy api would be better to use the zk one. Search the code for archived_root. I think the newer rkyv uses `access`